pkg/chartsync/git.go

```diff
@@ -0,0 +1,337 @@
+// updates for `v1.HelmRelease`s the sync changes are relevant for.
+type GitChartSync struct {
+	logger log.Logger
+	config GitConfig
+
+	lister lister.HelmReleaseLister
+
+	mirrors *git.Mirrors
+
+	sourcesMu sync.RWMutex
+	sources   map[string]sourceRef
+
+	releaseQueue ReleaseQueue
+}
+
+// sourceRef is used for book keeping, so that we know when a
+// signal we receive from a mirror is an actual update for a
+// release, and if the source we hold is still the one referred
+// to in the `v1.HelmRelease`.
+type sourceRef struct {
+	mirror string
+	remote string
+	ref    string
+	head   string
+}
+
+// forHelmRelease returns true if the given `v1.HelmRelease`s
+// `v1.GitChartSource` matches the sourceRef.
+func (c sourceRef) forHelmRelease(hr *v1.HelmRelease) bool {
+	if hr == nil || hr.Spec.GitChartSource == nil {
+		return false
+	}
+	return c.mirror == mirrorName(hr) && c.remote == hr.Spec.GitURL && c.ref == hr.Spec.Ref
+}
+
+func NewGitChartSync(logger log.Logger,
+	lister lister.HelmReleaseLister, cfg GitConfig, queue ReleaseQueue) *GitChartSync {
+
+	return &GitChartSync{
+		logger:       logger,
+		config:       cfg,
+
+		lister:       lister,
+
+		mirrors:      git.NewMirrors(),
+		sources:      make(map[string]sourceRef),
+
+		releaseQueue: queue,
+	}
+}
+
+// Run starts the mirroring of git repositories, and processes mirror
+// changes on signal, scheduling a release for a `HelmRelease` resource
+// when the update is relevant to the release.
+func (c *GitChartSync) Run(stopCh <-chan struct{}, errCh chan error, wg *sync.WaitGroup) {
+	c.logger.Log("info", "starting sync of git chart sources")
+
+	wg.Add(1)
+	go func() {
+		defer func() {
+			c.mirrors.StopAllAndWait()
+			wg.Done()
+		}()
+
+		for {
+			select {
+			case changed := <-c.mirrors.Changes():
+				for mirrorName := range changed {
+					repo, ok := c.mirrors.Get(mirrorName)
+
+					hrs, err := c.helmReleasesForMirror(mirrorName)
+					if err != nil {
+						c.logger.Log("error", ErrReleasesForMirror.Error(), "mirror", mirrorName, "err", err)
+						continue
+					}
+
+					// We received a signal from a no longer existing
+					// mirror.
+					if !ok {
+						if len(hrs) == 0 {
+							// If there are no references to it either,
+							// just continue with the next mirror...
+							continue
+						}
+
+						c.logger.Log("warning", ErrNoMirror.Error(), "mirror", mirrorName)
+						for _, hr := range hrs {
+							c.maybeMirror(mirrorName, hr.Spec.GitChartSource.GitURL)
+						}
+						// Wait for the signal from the newly requested mirror...
+						continue
+					}
+
+					c.processChangedMirror(mirrorName, repo, hrs)
+				}
+			case <-stopCh:
+				c.logger.Log("info", "stopping sync of git chart sources")
+				return
+			}
+		}
+	}()
+}
+
+// ChartDir returns a newly exported `GitChartDir` for the given `v1.HelmRelease`
+// at the latest HEAD, or an error.
+func (c *GitChartSync) ChartDir(hr *v1.HelmRelease) (GitChartDir, error) {
```

Comment:
I would rename this to something with a verb (since it requires some work) and more descriptive. Something like `GetChartSnapshotDir()` or `GetMirrorCopy()` or `TakeChartMirrorSnapshot()` or `TakeChartSnapshot()`